### DISCUSSION TITLE
sqlalchemy 2.0 in a long running app: Session.begin() vs Session w/ commit-as-you-go

### DISCUSSION THREAD

**Discussion by jreich218** [sqlalchemy/sqlalchemy#7175] [Usage Questions]
I'm building a long running app (basically a python app in an infinite loop) that does many db transactions per iteration that each need to be committed right after the transaction does its last DML statement. 

For whatever its worth, each iteration starts w/ fresh "everything"... so no db state is needed from any object from the previous loop. 

Basically a normal long running app in Python... no Flask/no web server etc)

Given that we want to have each transaction in a context provider, is it better to:
1. Use sessionmaker with Session.begin() every time a new transaction starts. The engine factory is used at the very beginning of the app before the infinite loop, so that would only run once... but, even still, is it going to slow the app down to have so many sessions starting up all the time?
2. Use sessionmaker with Session.begin() just before the infinite loop starts so that there is just one long-term session... and then use commit-as-you-go for each transaction using perhaps a custom made context manager that uses the session connection for: "begin, try, db stuff, except, rollback/raise, else commit"?
3. Some other pattern I don't know about that is meant for this situation?

Any feedback would be great. Thanks...

**Comment by CaselIT** [comment]
Hi,

If each iteration of the loop should be its own transaction and is independent from the other iteration I would create a new session for each iteration, similarly to how the suggested pattern of having a new session for each request in a webserver.

Not sure if I answered your question.
